wsgi.py

```diff
@@ -5,13 +5,28 @@
 import tempfile
 import json
 import re
+import sys
 
 from flask import Flask, jsonify, request
 from flask.logging import default_handler
 import requests
 
 from publish_json_schema import PublishJSONSchema
-import prometheus_metrics
+
+# all gunicorn processes in a given instance need to access a common
+# folder in /tmp where the metrics can be recorded
+PROMETHEUS_MULTIPROC_DIR = '/tmp/aiops_publisher'
+
+try:
+    os.makedirs(PROMETHEUS_MULTIPROC_DIR, exist_ok=True)
+    os.environ['prometheus_multiproc_dir'] = PROMETHEUS_MULTIPROC_DIR
+    import prometheus_metrics
+except IOError as e:
+    # this is a non-starter for scraping metrics in the
+    # Multiprocess Mode (Gunicorn)
+    # terminate if there is an exception here
+    sys.exit("Error while creating prometheus_multiproc_dir: " + str(e))
```

Comment:
Why are we exiting here instead of raising an exception? Honest question, not sure what's the difference in behavior in OpenShift. Raising an exception is at least caught in logs...

Reply:
If there is an exception while creating that that folder in `/tmp`, then Metrics multiprocessor collection will not work, which means M&M for this microservice will either not work at all or will continue reporting wrong results.
So just terminate, since the app is now incapable of scraping metrics

This is what I noticed in OpenShift after artificially simulating an error with -
```diff
diff --git a/wsgi.py b/wsgi.py
index efb87c7..e8d842f 100644
--- a/wsgi.py
+++ b/wsgi.py
@@ -15,10 +15,10 @@ from publish_json_schema import PublishJSONSchema
 
 # all gunicorn processes in a given instance need to access a common
 # folder in /tmp where the metrics can be recorded
-PROMETHEUS_MULTIPROC_DIR = '/tmp/aiops_publisher'
+PROMETHEUS_MULTIPROC_DIR = '/tmp'
 
 try:
-    os.makedirs(PROMETHEUS_MULTIPROC_DIR, exist_ok=True)
+    os.makedirs(PROMETHEUS_MULTIPROC_DIR, exist_ok=False)
     os.environ['prometheus_multiproc_dir'] = PROMETHEUS_MULTIPROC_DIR
     import prometheus_metrics
 except IOError as e:
```

OpenShift log -
```
---> Serving application with gunicorn (wsgi) ...
  | [2019-01-28 21:47:54 +0000] [1] [INFO] Starting gunicorn 19.9.0
  | [2019-01-28 21:47:54 +0000] [1] [INFO] Listening at: http://0.0.0.0:8080  (1)
  | [2019-01-28 21:47:54 +0000] [1] [INFO] Using worker: sync
  | [2019-01-28 21:47:54 +0000] [34] [INFO] Booting worker with pid: 34
  | [2019-01-28 21:47:54 +0000] [35] [INFO] Booting worker with pid: 35
  | [2019-01-28 21:47:54 +0000] [36] [INFO] Booting worker with pid: 36
  | Error while creating prometheus_multiproc_dir: [Errno 17] File exists: '/tmp'
  | [2019-01-28 21:47:55 +0000] [35] [INFO] Worker exiting (pid: 35)
  | Error while creating prometheus_multiproc_dir: [Errno 17] File exists: '/tmp'
  | [2019-01-28 21:47:55 +0000] [34] [INFO] Worker exiting (pid: 34)
  | Error while creating prometheus_multiproc_dir: [Errno 17] File exists: '/tmp'
  | [2019-01-28 21:47:55 +0000] [36] [INFO] Worker exiting (pid: 36)
  | [2019-01-28 21:47:55 +0000] [40] [INFO] Booting worker with pid: 40
  | [2019-01-28 21:47:55 +0000] [41] [INFO] Booting worker with pid: 41
  | [2019-01-28 21:47:55 +0000] [42] [INFO] Booting worker with pid: 42


```
. . . and the cycle continues
So looks like it terminates the worker, and spawns a new one

Reply:
Hm.. I mean, I understand the reasoning of the check. And it's a good check, no doubt. :wink:

My question is if raising an exception wouldn't be better solution than `sys.exit` call to avoid this looping. Because here's what I've found. Let's run a python script via gunicorn. This script doesn't do anything, it just wants to shut down the server:

### Testing `sys.exit()`
`sys_exit.py`:
```python
import sys

print('We should exit now!')
sys.exit("Exiting by sys.exit call")
```
Run it:
```bash
$ gunicorn sys_exit -w 3
[2019-01-29 08:42:26 +0100] [22686] [INFO] Starting gunicorn 19.9.0
[2019-01-29 08:42:26 +0100] [22686] [INFO] Listening at: http://0.0.0.0:8001 (22686)
[2019-01-29 08:42:26 +0100] [22686] [INFO] Using worker: sync
[2019-01-29 08:42:26 +0100] [22689] [INFO] Booting worker with pid: 22689
We should exit now!
[2019-01-29 08:42:26 +0100] [22689] [INFO] Worker exiting (pid: 22689)
Exiting by sys.exit call
[2019-01-29 08:42:26 +0100] [22690] [INFO] Booting worker with pid: 22690
We should exit now!
[2019-01-29 08:42:26 +0100] [22690] [INFO] Worker exiting (pid: 22690)
Exiting by sys.exit call
[2019-01-29 08:42:27 +0100] [22691] [INFO] Booting worker with pid: 22691
We should exit now!
[2019-01-29 08:42:27 +0100] [22691] [INFO] Worker exiting (pid: 22691)
Exiting by sys.exit call
...
```
:arrow_right: **This doesn't work**, we're looping and spawning new workers every time one worker fails.

### Testing `raise RuntimeError`
`raise.py`:
```python
print('We should exit now!')
raise RuntimeError("Exiting by direct exception raise")
```
Run it:
```bash
$ gunicorn raise -w 3
[2019-01-29 08:47:23 +0100] [24568] [INFO] Starting gunicorn 19.9.0
[2019-01-29 08:47:23 +0100] [24568] [INFO] Listening at: http://0.0.0.0:8001 (24568)
[2019-01-29 08:47:23 +0100] [24568] [INFO] Using worker: sync
[2019-01-29 08:47:23 +0100] [24571] [INFO] Booting worker with pid: 24571
We should exit now!
[2019-01-29 08:47:23 +0100] [24571] [ERROR] Exception in worker process
Traceback (most recent call last):
  File "/home/tcoufal/.local/share/virtualenvs/data-collector-5JyMlcdN/lib/python3.6/site-packages/gunicorn/arbiter.py", line 583, in spawn_worker
    worker.init_process()
  File "/home/tcoufal/.local/share/virtualenvs/data-collector-5JyMlcdN/lib/python3.6/site-packages/gunicorn/workers/base.py", line 129, in init_process
    self.load_wsgi()
  File "/home/tcoufal/.local/share/virtualenvs/data-collector-5JyMlcdN/lib/python3.6/site-packages/gunicorn/workers/base.py", line 138, in load_wsgi
    self.wsgi = self.app.wsgi()
  File "/home/tcoufal/.local/share/virtualenvs/data-collector-5JyMlcdN/lib/python3.6/site-packages/gunicorn/app/base.py", line 67, in wsgi
    self.callable = self.load()
  File "/home/tcoufal/.local/share/virtualenvs/data-collector-5JyMlcdN/lib/python3.6/site-packages/gunicorn/app/wsgiapp.py", line 52, in load
    return self.load_wsgiapp()
  File "/home/tcoufal/.local/share/virtualenvs/data-collector-5JyMlcdN/lib/python3.6/site-packages/gunicorn/app/wsgiapp.py", line 41, in load_wsgiapp
    return util.import_app(self.app_uri)
  File "/home/tcoufal/.local/share/virtualenvs/data-collector-5JyMlcdN/lib/python3.6/site-packages/gunicorn/util.py", line 350, in import_app
    __import__(module)
  File "/home/tcoufal/Programming/AI-Ops/data-collector/x.py", line 6, in <module>
    raise RuntimeError("Exiting by direct exception raise")
RuntimeError: Exiting by direct exception raise
[2019-01-29 08:47:23 +0100] [24571] [INFO] Worker exiting (pid: 24571)
[2019-01-29 08:47:23 +0100] [24568] [INFO] Shutting down: Master
[2019-01-29 08:47:23 +0100] [24568] [INFO] Reason: Worker failed to boot.
```
:arrow_right: **Works!** Raising an exception can stop gunicorn to spawn new workers.

### Even better solution: `Arbiter.APP_LOAD_ERROR`

Digging deep in Gunicorn, I've found their "emergency shutdown" mechanism. [A special error code](https://github.com/benoitc/gunicorn/blob/9f87c88819b901d4cd8ddf6c2bba1198b68903cb/gunicorn/arbiter.py#L34) which indicates something went horribly wrong during the start up. Let's test this one:

`emergency_brake.py`:
```python
import sys
from gunicorn.arbiter import Arbiter

print('We should exit now!')
sys.exit(Arbiter.APP_LOAD_ERROR)
```

Let's run it:
```bash
$ gunicorn emergency_brake -w 3
[2019-01-29 08:57:52 +0100] [26032] [INFO] Starting gunicorn 19.9.0
[2019-01-29 08:57:52 +0100] [26032] [INFO] Listening at: http://0.0.0.0:8001 (26032)
[2019-01-29 08:57:52 +0100] [26032] [INFO] Using worker: sync
[2019-01-29 08:57:52 +0100] [26035] [INFO] Booting worker with pid: 26035
We should exit now!
[2019-01-29 08:57:52 +0100] [26035] [INFO] Worker exiting (pid: 26035)
[2019-01-29 08:57:52 +0100] [26032] [INFO] Shutting down: Master
[2019-01-29 08:57:52 +0100] [26032] [INFO] Reason: App failed to load.
```

:arrow_right: **Works!** It's a cleaner solution then the second option and it uses a native mechanism...

Both later two solutions can satisfy our needs, the first doesn't work for me. Does it for you? That's what've meant by my question. The looping is something we both know we should avoid, because it's much cleaner to let the container fail and die... (Sad but true, lol.)

Note: Instead of the `print` I've used in my examples it would be nice to use `logger.error` in the real code.

Reply:
Thanks for all the options presented above and the details!

Let's go with `Arbiter.APP_LOAD_ERROR` - easy to implement, although I might have to move the logger definition up...but that should be OK

> it's much cleaner to let the container fail and die... (Sad but true, lol.)

Agree that this is better than resurrecting the container again :)

Reply:
Don't worry about the logging config. Just move it around as you wish, it's fine. :wink: 

I agree, the third option sounds the best to me as well. It's elegant, and the log output is neat as well. Let's do this! :+1: